### ISSUE TITLE
Variable REQUIRE_BASE_URL for standalone modules?

### ISSUE THREAD

**Issue by karolyi** [etianen/django-require#58]
Hi,

I am trying to compile standalone modules residing in different subdirectories, each being a different subsystem, independent of each other.

Since each of the standalone modules needs its own configuration, the `REQUIRE_BASE_URL` should be different for them, or else the compilation part will fail, telling that it can't find files that are actually in subdirectories. But since I can't change the `baseUrl` to their respective subdirectories, the build will fail.

This could be fixed with introducing a separate `BASE_URL` setting to each `REQUIRE_STANDALONE_MODULES` section, which can be relative to the `REQUIRE_BASE_URL` setting.

Could you please implement this?

**Comment by etianen** [comment]
That sounds a plausible use-case, and I'd be happy to take a pull request implementing it. The feature would presumably fall back to the global REQUIRE_BASE_URL if not specified.

**Comment by karolyi** [comment]
Thank you for your answer!

Since what I was imagining breaks the existing functionality, for backwards compatibility I'd keep the existing tags and create new tags using partially the same functionality, so the module wouldn't break for people already using it. The new tags would use the 'extended' configuration.

If you're okay with that, I'd be happy to create a pull request eventually.

**Comment by etianen** [comment]
Why does this break existing functionality? So long as you don't specify a
base_url setting for a standalone module, it will keep working as before.

**Comment by karolyi** [comment]
Yes, I thought you would ask that, so in order to explain, let me introduce my usecase (I was never the theoretical guy:) ):

I have an application, for which I plan to have more skins, that is templates, css (SASS) files, javascript files. So I have the following directory structure for the static files:

```
static/
-- bower_components/
-- skins/
---- app.build.js
---- default/
------ scss/
------ js/
-------- common.js
-------- common.build.js
-------- sidebar.js and other JS modules for the skin
---- dark/
------ scss/
------ js/
-------- common.js
-------- common.build.js
-------- other JS modules for the skin
```

So as you see, I use the `common.js` for each skin as an entry point, and then load the widget modules for the respective widgets inlined (so I can pass them some dynamic configuration).

I have the following config for `django-require`:
```python
REQUIRE_BASE_URL = 'skins'
REQUIRE_JS = '../bower_components/requirejs/require.js'
REQUIRE_BUILD_PROFILE = 'app.build.js'
REQUIRE_DEBUG = DEBUG
REQUIRE_STANDALONE_MODULES = {
    'default/js/common': {
        'out': 'default/js/common-built.js',
        'build_profile': 'default/js/common.build.js',
    }
}
```

which works in development (`DEBUG = True`) mode, with using `{% require_module 'default/js/common' %}`, but fails when I try to build the modules. The reason is the wrong baseUrl specification. While using it in development mode, the tag produces the following output:
```html
<script src="/static/bower_components/requirejs/require.js" data-main="/static/skins/default/js/common.js"></script>
```

Which means, RequireJS will set its `baseUrl` to the module path, so everything would work just fine when importing the sidebar module, for example:
```html
<script type="text/javascript">
    require(['sidebar'], function (sidebar) {
      sidebar.init({
        selectors: {
            buttonToggle: '.navbar-toggle',
            sidebarWrapper: '.sidebar-wrapper',
            sidebarOverlay: '.sidebar-offcanvas-overlay'
        }
      });
    });
</script>
```
But fails to build when doing `collectstatic`, as the `baseUrl` setting is overridden there: adding `sidebar.js` to the included modules in `common.build.js` will result the compilation failing because of searching for that file under `skins/`.

What I'd like to achieve is to have each compiled module placed under `/static/skins/<skinname>/js/`, that's where the relative` BASE_URL` comes in. So I started tweaking your code, adding the relative `BASE_URL`, and came to the conclusion that `almond.js` has to be copied in the same directory level. That's okay, but another problem occurs: since I use `'default/js/common'` as the modules name, which has a path in it in order to use it in the HTML properly, the introduced relative path will break the filename. So here I'd like to introduce a filename next to the relative path, while using the config key only as a property key, rather than a something that refers to a filename.

---------

The second thing I'd like to achieve is to make sure that `common.js` (which contains the require.config AND the basic module definition) gets loaded before all modules in development mode. RequireJS advises to use two `<script>` tags in the header part for that, one for loading `require.js` itself, and then the other to load `common.js`. This would load these two scripts before everything else, since they would be in the header at the top level. The point of this is, this way I won't need to require `common.js` in each inline require tag, as suggested [here](http://requirejs.org/docs/api.html#data-main).

So in development mode, my new tag would produce a tag something like this:
```html
<script src="/static/bower_components/requirejs/require.js"></script>
<script src="/static/skins/default/js/common.js"></script>
```
Of course for this approach, I'd have to set `baseUrl` in the config, but that's something I can live with. Since at `collectstatic` I'd compile everything together in one file with almond, this wouldn't be an issue there.

------------------

To sum up:
- Adding entry point filenames and relative paths to the currently working tag would break the existing functionality for most of the people
- The new tag would produce a slightly different HTML output in development mode, for which a different approach is needed.

I'm really open for a discussion on this, but keep in mind I've been playing around with this in the last several days trying to find a way, and to me this seems the only plausible solution.

**Comment by etianen** [comment]
I think that your problem could be easily solved by placing all the entry
points for your skins in the REQUIRE_BASE_URL directory. This is something
RequireJS generally recommends, as it makes the configuration much easier.
Something like:


   - js/
      - default.js
      - default/
         - scss/
         - other-default-skin-file.js
      - dark.js
      - dark/
         - scss/
         - other-dark-skin-file.js

It's a bit less neat, but it should solve your issues with unduly
complicating django-require.

**Comment by karolyi** [comment]
I was making thoughts of before, but besides the 'misplaced' files, there's another problem:

you should refer all your modules starting from the require_base directory, which I find an unnecessary complication, since I could configure the proper baseUrl. In your version, I should require sidebar.js like as "require(['/js/default/sidebar'], function)".

**Comment by etianen** [comment]
Here's the thing though... it took you a massive email to explain your
issue, and I still don't full understand it. Explaining that to other users
with variegated use cases will be just as difficult. The general rule of
"place entry points in the root" is much easier to grok.

**Comment by karolyi** [comment]
so you say we leave it?

**Comment by etianen** [comment]
I think so, in this case. I'm reluctant to add a lot of extra complexity to
support a complex use-case, when a simpler use-case will achieve the same
thing.

RequireJS is a rabbit-hole of baffling configuration options, most of which
can be avoided by keeping it simple.

**Comment by karolyi** [comment]
I definitely agree with you on the RequireJS issue.

If that's the case, would you have anything against if I created a separate repository out of your code, and implement it there for myself?

**Comment by etianen** [comment]
Not at all. It's open source, so feel free to clone it on GitHub and for
your thing. :)

**Comment by karolyi** [comment]
Thanks.

If you're interested I'll keep you updated.

**Comment by etianen** [comment]
Please do! If it turns out to be a simple implementation with few side
effects, I'd still consider merging.

**Comment by karolyi** [comment]
hey,

Just checking in, I wanted to let you know that I'm more or less ready with the modifications. Your project got a huuuge overhaul and upgrade, refactoring, code coverage, automated tests via tox and whatnot :) Check it in my repo.

Before submitting a pull request, I'll test it with my own project and also update the documentation.

**Comment by etianen** [comment]
Hi! Thanks for letting me know. :)

I'm afraid, though, that there's no real way that I can just pull in a complete overhaul of the project in one big lump. The responsibility for maintaining this repo means that I need an understanding of the code, and there's now 2.5K lines of code that I didn't write, all turning up at once.

I'm happy to merge self-contained pull requests that have a clear aim and provide improvements in terms of new features, better test coverage, fixing bugs etc. A single massive refactor isn't something I can support in the long-term, however.

**Comment by karolyi** [comment]
It's not such a huge overhaul. If you look into the code I'll make a pull request out of, it's basically just refactorings and add a bunch of new functionality here and there.

Especially, the huge `post_process` function was dissected to give a it more clear understanding. The code is definitely not broken, since all the unittests run successfully (build is green). Also, in order to achieve 100% code coverage, I started out with adding tests to the original code, so I can refactor stuff without breaking the build. Meanwhile I introduced tox/travis/coverage, so everything is more automated now.

It's just that I updated its requirements to todays needs.

I have a suggestion: maybe you could create an organization/add me as a repo maintainer, so I could help you out with the maintenance.

Cheers,
--
László Károlyi
http://linkedin.com/in/karolyi

**Comment by karolyi** [comment]
[Done.](https://github.com/karolyi/django-require2)
